# SHP2_alignment_motif_mapping_annotracks

This repository contains Python scripts that convert Dialign-Chaos[1] multiple sequence alignment output, and PatMan[2] mapping output to annotation tracks that are compatible with the GBrowse tool available on The Arabidopsis Information Resource (TAIR) for the TAIR10 version of the Arabidopsis thaliana genome.

Ref:

[1] Brudno, M., Steinkamp, R., and Morgenstern, B. (2004). The CHAOS/DIALIGN WWW server for multiple alignment of genomic sequences. Nucleic Acids Res. 32: W41–4.

[2]Prüfer, K., Stenzel, U., Dannemann, M., Green, R.E., Lachmann, M., and Kelso, J. (2008). PatMaN: rapid alignment of short sequences to large databases. Bioinformatics 24: 1530–1531.
